File: binding/include/libqmdiarea.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmdiarea.html
pub const qmdiarea = struct {

    /// New constructs a new QMdiArea object.
    pub fn New(parent: ?*anyopaque) QtC.QMdiArea {
        return qtc.QMdiArea_new(@ptrCast(parent));
    }


    /// New2 constructs a new QMdiArea object.
    pub fn New2() QtC.QMdiArea {
        return qtc.QMdiArea_new2();
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMdiArea_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmdiarea.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QMdiArea_SizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QMdiArea_OnSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QMdiArea_QBaseSizeHint(@ptrCast(self));
    }

    pub fn MinimumSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QMdiArea_MinimumSizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnMinimumSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QMdiArea_OnMinimumSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMinimumSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QMdiArea_QBaseMinimumSizeHint(@ptrCast(self));
    }

    pub fn CurrentSubWindow(self: ?*anyopaque, ) QtC.QMdiSubWindow {
        return qtc.QMdiArea_CurrentSubWindow(@ptrCast(self));
    }

    pub fn ActiveSubWindow(self: ?*anyopaque, ) QtC.QMdiSubWindow {
        return qtc.QMdiArea_ActiveSubWindow(@ptrCast(self));
    }

    pub fn SubWindowList(self: ?*anyopaque, ) []const u8 {
        return qtc.QMdiArea_SubWindowList(@ptrCast(self));
    }

    pub fn AddSubWindow(self: ?*anyopaque, widget: ?*anyopaque) QtC.QMdiSubWindow {
        return qtc.QMdiArea_AddSubWindow(@ptrCast(self), @ptrCast(widget));
    }

    pub fn RemoveSubWindow(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QMdiArea_RemoveSubWindow(@ptrCast(self), @ptrCast(widget));
    }

    pub fn Background(self: ?*anyopaque, ) QtC.QBrush {
        return qtc.QMdiArea_Background(@ptrCast(self));
    }

    pub fn SetBackground(self: ?*anyopaque, background: ?*anyopaque) void {
        qtc.QMdiArea_SetBackground(@ptrCast(self), @ptrCast(background));
    }

    pub fn ActivationOrder(self: ?*anyopaque, ) i32 {
        return qtc.QMdiArea_ActivationOrder(@ptrCast(self));
    }

    pub fn SetActivationOrder(self: ?*anyopaque, order: i32) void {
        qtc.QMdiArea_SetActivationOrder(@ptrCast(self), @intCast(order));
    }

    pub fn SetOption(self: ?*anyopaque, option: i32) void {
        qtc.QMdiArea_SetOption(@ptrCast(self), @intCast(option));
    }

    pub fn TestOption(self: ?*anyopaque, opton: i32) bool {
        return qtc.QMdiArea_TestOption(@ptrCast(self), @intCast(opton));
    }

    pub fn SetViewMode(self: ?*anyopaque, mode: i32) void {
        qtc.QMdiArea_SetViewMode(@ptrCast(self), @intCast(mode));
    }

    pub fn ViewMode(self: ?*anyopaque, ) i32 {
        return qtc.QMdiArea_ViewMode(@ptrCast(self));
    }

    pub fn DocumentMode(self: ?*anyopaque, ) bool {
        return qtc.QMdiArea_DocumentMode(@ptrCast(self));
    }

    pub fn SetDocumentMode(self: ?*anyopaque, enabled: bool) void {
        qtc.QMdiArea_SetDocumentMode(@ptrCast(self), enabled);
    }

    pub fn SetTabsClosable(self: ?*anyopaque, closable: bool) void {
        qtc.QMdiArea_SetTabsClosable(@ptrCast(self), closable);
    }

    pub fn TabsClosable(self: ?*anyopaque, ) bool {
        return qtc.QMdiArea_TabsClosable(@ptrCast(self));
    }

    pub fn SetTabsMovable(self: ?*anyopaque, movable: bool) void {
        qtc.QMdiArea_SetTabsMovable(@ptrCast(self), movable);
    }

    pub fn TabsMovable(self: ?*anyopaque, ) bool {
        return qtc.QMdiArea_TabsMovable(@ptrCast(self));
    }

    pub fn SetTabShape(self: ?*anyopaque, shape: i32) void {
        qtc.QMdiArea_SetTabShape(@ptrCast(self), @intCast(shape));
    }

    pub fn TabShape(self: ?*anyopaque, ) i32 {
        return qtc.QMdiArea_TabShape(@ptrCast(self));
    }

    pub fn SetTabPosition(self: ?*anyopaque, position: i32) void {
        qtc.QMdiArea_SetTabPosition(@ptrCast(self), @intCast(position));
    }

    pub fn TabPosition(self: ?*anyopaque, ) i32 {
        return qtc.QMdiArea_TabPosition(@ptrCast(self));
    }

    pub fn SubWindowActivated(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QMdiArea_SubWindowActivated(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OnSubWindowActivated(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_Connect_SubWindowActivated(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn SetActiveSubWindow(self: ?*anyopaque, window: ?*anyopaque) void {
        qtc.QMdiArea_SetActiveSubWindow(@ptrCast(self), @ptrCast(window));
    }

    pub fn TileSubWindows(self: ?*anyopaque, ) void {
        qtc.QMdiArea_TileSubWindows(@ptrCast(self));
    }

    pub fn CascadeSubWindows(self: ?*anyopaque, ) void {
        qtc.QMdiArea_CascadeSubWindows(@ptrCast(self));
    }

    pub fn CloseActiveSubWindow(self: ?*anyopaque, ) void {
        qtc.QMdiArea_CloseActiveSubWindow(@ptrCast(self));
    }

    pub fn CloseAllSubWindows(self: ?*anyopaque, ) void {
        qtc.QMdiArea_CloseAllSubWindows(@ptrCast(self));
    }

    pub fn ActivateNextSubWindow(self: ?*anyopaque, ) void {
        qtc.QMdiArea_ActivateNextSubWindow(@ptrCast(self));
    }

    pub fn ActivatePreviousSubWindow(self: ?*anyopaque, ) void {
        qtc.QMdiArea_ActivatePreviousSubWindow(@ptrCast(self));
    }

    pub fn SetupViewport(self: ?*anyopaque, viewport: ?*anyopaque) void {
        qtc.QMdiArea_SetupViewport(@ptrCast(self), @ptrCast(viewport));
    }

    /// Allows for overriding the related default method
    pub fn OnSetupViewport(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_OnSetupViewport(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetupViewport(self: ?*anyopaque, viewport: ?*anyopaque) void {
        qtc.QMdiArea_QBaseSetupViewport(@ptrCast(self), @ptrCast(viewport));
    }

    pub fn Event(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMdiArea_Event(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QMdiArea_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMdiArea_QBaseEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn EventFilter(self: ?*anyopaque, object: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMdiArea_EventFilter(@ptrCast(self), @ptrCast(object), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEventFilter(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QMdiArea_OnEventFilter(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEventFilter(self: ?*anyopaque, object: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMdiArea_QBaseEventFilter(@ptrCast(self), @ptrCast(object), @ptrCast(event));
    }

    pub fn PaintEvent(self: ?*anyopaque, paintEvent: ?*anyopaque) void {
        qtc.QMdiArea_PaintEvent(@ptrCast(self), @ptrCast(paintEvent));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_OnPaintEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEvent(self: ?*anyopaque, paintEvent: ?*anyopaque) void {
        qtc.QMdiArea_QBasePaintEvent(@ptrCast(self), @ptrCast(paintEvent));
    }

    pub fn ChildEvent(self: ?*anyopaque, childEvent: ?*anyopaque) void {
        qtc.QMdiArea_ChildEvent(@ptrCast(self), @ptrCast(childEvent));
    }

    /// Allows for overriding the related default method
    pub fn OnChildEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_OnChildEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseChildEvent(self: ?*anyopaque, childEvent: ?*anyopaque) void {
        qtc.QMdiArea_QBaseChildEvent(@ptrCast(self), @ptrCast(childEvent));
    }

    pub fn ResizeEvent(self: ?*anyopaque, resizeEvent: ?*anyopaque) void {
        qtc.QMdiArea_ResizeEvent(@ptrCast(self), @ptrCast(resizeEvent));
    }

    /// Allows for overriding the related default method
    pub fn OnResizeEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_OnResizeEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseResizeEvent(self: ?*anyopaque, resizeEvent: ?*anyopaque) void {
        qtc.QMdiArea_QBaseResizeEvent(@ptrCast(self), @ptrCast(resizeEvent));
    }

    pub fn TimerEvent(self: ?*anyopaque, timerEvent: ?*anyopaque) void {
        qtc.QMdiArea_TimerEvent(@ptrCast(self), @ptrCast(timerEvent));
    }

    /// Allows for overriding the related default method
    pub fn OnTimerEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_OnTimerEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseTimerEvent(self: ?*anyopaque, timerEvent: ?*anyopaque) void {
        qtc.QMdiArea_QBaseTimerEvent(@ptrCast(self), @ptrCast(timerEvent));
    }

    pub fn ShowEvent(self: ?*anyopaque, showEvent: ?*anyopaque) void {
        qtc.QMdiArea_ShowEvent(@ptrCast(self), @ptrCast(showEvent));
    }

    /// Allows for overriding the related default method
    pub fn OnShowEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMdiArea_OnShowEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseShowEvent(self: ?*anyopaque, showEvent: ?*anyopaque) void {
        qtc.QMdiArea_QBaseShowEvent(@ptrCast(self), @ptrCast(showEvent));
    }

    pub fn ViewportEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMdiArea_ViewportEvent(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnViewportEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QMdiArea_OnViewportEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseViewportEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMdiArea_QBaseViewportEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn ScrollContentsBy(self: ?*anyopaque, dx: i32, dy: i32) void {
        qtc.QMdiArea_ScrollContentsBy(@ptrCast(self), dx, dy);
    }

    /// Allows for overriding the related default method
    pub fn OnScrollContentsBy(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, i32) callconv(.c) void) void {
        qtc.QMdiArea_OnScrollContentsBy(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseScrollContentsBy(self: ?*anyopaque, dx: i32, dy: i32) void {
        qtc.QMdiArea_QBaseScrollContentsBy(@ptrCast(self), dx, dy);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMdiArea_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmdiarea.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMdiArea_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmdiarea.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SubWindowList1(self: ?*anyopaque, order: i32) []const u8 {
        return qtc.QMdiArea_SubWindowList1(@ptrCast(self), @intCast(order));
    }

    pub fn AddSubWindow2(self: ?*anyopaque, widget: ?*anyopaque, flags: i64) QtC.QMdiSubWindow {
        return qtc.QMdiArea_AddSubWindow2(@ptrCast(self), @ptrCast(widget), @intCast(flags));
    }

    pub fn SetOption2(self: ?*anyopaque, option: i32, on: bool) void {
        qtc.QMdiArea_SetOption2(@ptrCast(self), @intCast(option), on);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMdiArea_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/mdiarea.html#types
pub const enums = struct {
    pub const AreaOption = enum {
        pub const DontMaximizeSubWindowOnActivation: i32 = 1;
    };

    pub const WindowOrder = enum {
        pub const CreationOrder: i32 = 0;
        pub const StackingOrder: i32 = 1;
        pub const ActivationHistoryOrder: i32 = 2;
    };

    pub const ViewMode = enum {
        pub const SubWindowView: i32 = 0;
        pub const TabbedView: i32 = 1;
    };

};

File: binding/src/libqmdiarea.cpp
#include <QAbstractScrollArea>
#include <QBrush>
#include <QChildEvent>
#include <QEvent>
#include <QMdiArea>
#include <QMdiSubWindow>
#include <QObject>
#include <QPaintEvent>
#include <QResizeEvent>
#include <QShowEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QTimerEvent>
#include <QWidget>
#include <qmdiarea.h>
#include "libqmdiarea.h"
#include "libqmdiarea.hxx"

QMdiArea* QMdiArea_new(QWidget* parent) {
	 return new VirtualQMdiArea(parent);
}

QMdiArea* QMdiArea_new2() {
	 return new VirtualQMdiArea();
}

libqt_string QMdiArea_Tr(const char* s) {
	QString _ret = QMdiArea::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QSize* QMdiArea_SizeHint(const QMdiArea* self) {
	return new QSize(self->sizeHint());
}

QSize* QMdiArea_MinimumSizeHint(const QMdiArea* self) {
	return new QSize(self->minimumSizeHint());
}

QMdiSubWindow* QMdiArea_CurrentSubWindow(const QMdiArea* self) {
	return self->currentSubWindow();
}

QMdiSubWindow* QMdiArea_ActiveSubWindow(const QMdiArea* self) {
	return self->activeSubWindow();
}

libqt_list QMdiArea_SubWindowList(const QMdiArea* self) {
	auto _ret = self->subWindowList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		_data[_i] = _ret[_i];
	}
	return _arr;
}

QMdiSubWindow* QMdiArea_AddSubWindow(QMdiArea* self, QWidget* widget) {
	return self->addSubWindow(widget);
}

void QMdiArea_RemoveSubWindow(QMdiArea* self, QWidget* widget) {
	self->removeSubWindow(widget);
}

QBrush* QMdiArea_Background(const QMdiArea* self) {
	return new QBrush(self->background());
}

void QMdiArea_SetBackground(QMdiArea* self, const QBrush* background) {
	self->setBackground(*background);
}

int QMdiArea_ActivationOrder(const QMdiArea* self) {
	return self->activationOrder();
}

void QMdiArea_SetActivationOrder(QMdiArea* self, int order) {
	self->setActivationOrder(static_cast<QMdiArea::WindowOrder>(order));
}

void QMdiArea_SetOption(QMdiArea* self, int option) {
	self->setOption(static_cast<QMdiArea::AreaOption>(option));
}

bool QMdiArea_TestOption(const QMdiArea* self, int opton) {
	return self->testOption(static_cast<QMdiArea::AreaOption>(opton));
}

void QMdiArea_SetViewMode(QMdiArea* self, int mode) {
	self->setViewMode(static_cast<QMdiArea::ViewMode>(mode));
}

int QMdiArea_ViewMode(const QMdiArea* self) {
	return self->viewMode();
}

bool QMdiArea_DocumentMode(const QMdiArea* self) {
	return self->documentMode();
}

void QMdiArea_SetDocumentMode(QMdiArea* self, bool enabled) {
	self->setDocumentMode(enabled);
}

void QMdiArea_SetTabsClosable(QMdiArea* self, bool closable) {
	self->setTabsClosable(closable);
}

bool QMdiArea_TabsClosable(const QMdiArea* self) {
	return self->tabsClosable();
}

void QMdiArea_SetTabsMovable(QMdiArea* self, bool movable) {
	self->setTabsMovable(movable);
}

bool QMdiArea_TabsMovable(const QMdiArea* self) {
	return self->tabsMovable();
}

void QMdiArea_SetTabShape(QMdiArea* self, int shape) {
	self->setTabShape(static_cast<QTabWidget::TabShape>(shape));
}

int QMdiArea_TabShape(const QMdiArea* self) {
	return self->tabShape();
}

void QMdiArea_SetTabPosition(QMdiArea* self, int position) {
	self->setTabPosition(static_cast<QTabWidget::TabPosition>(position));
}

int QMdiArea_TabPosition(const QMdiArea* self) {
	return self->tabPosition();
}

void QMdiArea_SubWindowActivated(QMdiArea* self, QMdiSubWindow* param1) {
	self->subWindowActivated(param1);
}

void QMdiArea_Connect_SubWindowActivated(QMdiArea* self, intptr_t slot) {
    void (*slotFunc)(QMdiArea*, QMdiSubWindow*) = reinterpret_cast<void (*)(QMdiArea*, QMdiSubWindow*)>(slot);
    QMdiArea::connect(self, &QMdiArea::subWindowActivated, [self, slotFunc](QMdiSubWindow* param1) {
	slotFunc(self, param1);
    });
}

void QMdiArea_SetActiveSubWindow(QMdiArea* self, QMdiSubWindow* window) {
	self->setActiveSubWindow(window);
}

void QMdiArea_TileSubWindows(QMdiArea* self) {
	self->tileSubWindows();
}

void QMdiArea_CascadeSubWindows(QMdiArea* self) {
	self->cascadeSubWindows();
}

void QMdiArea_CloseActiveSubWindow(QMdiArea* self) {
	self->closeActiveSubWindow();
}

void QMdiArea_CloseAllSubWindows(QMdiArea* self) {
	self->closeAllSubWindows();
}

void QMdiArea_ActivateNextSubWindow(QMdiArea* self) {
	self->activateNextSubWindow();
}

void QMdiArea_ActivatePreviousSubWindow(QMdiArea* self) {
	self->activatePreviousSubWindow();
}

libqt_string QMdiArea_Tr2(const char* s, const char* c) {
	QString _ret = QMdiArea::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMdiArea_Tr3(const char* s, const char* c, int n) {
	QString _ret = QMdiArea::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QMdiArea_SubWindowList1(const QMdiArea* self, int order) {
	auto _ret = self->subWindowList(static_cast<QMdiArea::WindowOrder>(order));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		_data[_i] = _ret[_i];
	}
	return _arr;
}

QMdiSubWindow* QMdiArea_AddSubWindow2(QMdiArea* self, QWidget* widget, int flags) {
	return self->addSubWindow(widget, static_cast<QFlags<Qt::WindowType>>(flags));
}

void QMdiArea_SetOption2(QMdiArea* self, int option, bool on) {
	self->setOption(static_cast<QMdiArea::AreaOption>(option), on);
}

// Base class handler implementation
QSize* QMdiArea_QBaseSizeHint(const QMdiArea* self) {
	auto* vqmdiarea = dynamic_cast<const VirtualQMdiArea*>(self);
	if (vqmdiarea && vqmdiarea->isVirtualQMdiArea) {
vqmdiarea->setQMdiArea_SizeHint_IsBase(true);
	return new QSize(vqmdiarea->sizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QMdiArea_OnSizeHint(const QMdiArea* self, intptr_t slot) {
	auto* vqmdiarea = dynamic_cast<const VirtualQMdiArea*>(self);
	if (vqmdiarea && vqmdiarea->isVirtualQMdiArea) {
vqmdiarea->setQMdiArea_SizeHint_Callback(reinterpret_cast<VirtualQMdiArea::QMdiArea_SizeHint_Callback>(slot));
}
}

// Base class handler implementation
QSize* QMdiArea_QBaseMinimumSizeHint(const QMdiArea* self) {
	auto* vqmdiarea = dynamic_cast<const VirtualQMdiArea*>(self);
	if (vqmdiarea && vqmdiarea->isVirtualQMdiArea) {
vqmdiarea->setQMdiArea_MinimumSizeHint_IsBase(true);
	return new QSize(vqmdiarea->minimumSizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QMdiArea_OnMinimumSizeHint(const QMdiArea* self, intptr_t slot) {
	auto* vqmdiarea = dynamic_cast<const VirtualQMdiArea*>(self);
	if (vqmdiarea && vqmdiarea->isVirtualQMdiArea) {
vqmdiarea->setQMdiArea_MinimumSizeHint_Callback(reinterpret_cast<VirtualQMdiArea::QMdiArea_MinimumSizeHint_Callback>(slot));
}
}

void QMdiArea_Delete(QMdiArea* self) {
    delete self;
}

